plugins/out_gelf/gelf.c

```diff
@@ -228,6 +229,51 @@ static int gelf_send_udp(struct flb_out_gelf_config *ctx, char *msg,
     return 0;
 }
 
+static int inject_tag(msgpack_object *map,
+                      struct flb_event_chunk *event_chunk,
+                      struct flb_out_gelf_config *ctx,
+                      char** out_buf, int* out_size)
+{
+    int i;
+    int len;
+    size_t map_num;
+    char *ret_buf;
+    msgpack_sbuffer sbuf;
+    msgpack_packer  pck;
+
+    len = map->via.map.size;
+    map_num = 1 + len;
+
+    msgpack_sbuffer_init(&sbuf);
+    msgpack_packer_init(&pck, &sbuf, msgpack_sbuffer_write);
+    msgpack_pack_map(&pck, map_num);
+
+    for (i = 0; i < len; i++) {
+        msgpack_pack_object(&pck, map->via.map.ptr[i].key);
+        msgpack_pack_object(&pck, map->via.map.ptr[i].val);
+    }
+
+    for (i = 0; i < 1; i++){
+        msgpack_pack_str(&pck, strlen(ctx->tag_key));
+        msgpack_pack_str_body(&pck, ctx->tag_key, strlen(ctx->tag_key));
+        msgpack_pack_str(&pck, flb_sds_len(event_chunk->tag));
+        msgpack_pack_str_body(&pck, event_chunk->tag, flb_sds_len(event_chunk->tag));
+    }
+
+    *out_size = sbuf.size;
+    ret_buf  = flb_malloc(sbuf.size);
+    *out_buf = ret_buf;
+    if (*out_buf == NULL) {
+        flb_errno();
+        msgpack_sbuffer_destroy(&sbuf);
+        return -1;
+    }
+    memcpy(*out_buf, sbuf.data, sbuf.size);
```

Comment:
here is possible to avoid the memory copy by simply assigning the pointer to the msgpack buffer, e.g:

```c
*out_buf = sbuf.data
*out_size = sbuf.size
```

just do not destroy sbuf before returning and should be fine

Reply:
Ah, got it. Direct operation should be fine if avoiding memory copy:

```c
static inline void msgpack_sbuffer_init(msgpack_sbuffer* sbuf)
{
    memset(sbuf, 0, sizeof(msgpack_sbuffer));
}

static inline void msgpack_sbuffer_destroy(msgpack_sbuffer* sbuf)
{
    free(sbuf->data);
}
```